### DISCUSSION TITLE
Undo/Redo Functionality

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#270] [Q&A]
There are a number of places in the controls where users might want to be able to undo/redo the actions that they take. This discussion is for two things:

- What kinds of things should plugin authors, flight directors, and crew members be able to undo/redo?
- What happens when two people are working on the same thing and one of them hits undo after the other makes a change to the thing that the undo is undoing?

I would love any thoughts you have about this!

**Comment by The3gs** [comment]
The main reason we want undo-redo is for editing, but I think that it would be worth it to think about where else it might be useful as well.

For editing of plugins we will not often have to deal with the second consideration, though I think it would be better to design it to be robust in such situations, rather than have it cause problems down the road. I think that it might be acceptable to just have each client just keep their own stacks and if an undo is attempted that would not longer be valid you could just have the server either ignore the request or send an error code back to the requesting client.

In regards to editing, we also will need to consider whether we would have a single undo/redo stack pair for each client, or if each page (Star map, Ship map, etc.) would have it's own undo/redo stacks. Points in favor of a single pair might be that it would be simpler to implement. I do think in some cases it would be unintuitive if you pressed undo on page A, and it undid an action that you had just done on page B (Which you might not even notice as you are not looking at that page.) I think that if we find it feasible we ought to have separate stacks for each page. To some extent, even having undo/redo is just for user experience, so I think we should try to have it be a intuitive as is possible.
(Note that many programs such as vscode have an independent stack for each file they have open, and in a lot of ways that is similar to the different screens we have in Thorium)

If the flight director and/or crew have undo/redo capability the only actions that should be undoable are things that can be undone manually, such as assigning a security team, setting energy allocations, or other things like those. We obviously would not have it so that they could undo an attack, as that is external to the ship's systems.

I am not familiar enough with the flight director's viewpoint to know what might be good for them to be able to undo.

**Reply by alexanderson1993** [reply]
According to my Flight Director sister, here are a few things flight directors could undo:
- Sending processed sensor data
- Sending long range messages
- Destroying a probe

Some of these might be more complicated to implement than I expected. For example, to destroy a probe, the client needs to store all of the data about all of the things on the probe and needs to have a way to tell the server to re-create that probe, but without triggering any of the side-effects. Normally when you create a probe it plays a sound effect and triggers some notifications, but undoing shouldn't do that.

This feature definitely won't be trivial to implement for in-flight situations, so maybe we limit it to the plugin editor initially.
